BZ-43226: feat: Add selective MCP server tool registration feature

BoraYaswanthReddy opened the pull request:
- Added register_selective_tools method to MCPClient for registering specific MCP tools
- Updated __init__.py to conditionally use selective tool registration

Dev proof
<img width="1434" height="619" alt="Screenshot 2025-08-11 at 3 41 41 PM" src="https://github.com/user-attachments/assets/1c3f51a8-374a-451c-be33-60161b10c885" />

While testing this, the selective_mcp_functions was
SELECTIVE_MCP_FUNCTIONS="shipping-rules-server_UpdateShippingZone,function2,shopify-analytics-server_getLowStockProducts,function3,breeze-analytics_TrafficSourceAttribution,utility-server_getCurrentTime"

BoraYaswanthReddy commented on app/agents/voice/automatic/__init__.py:
```diff
@@ -138,20 +153,29 @@ async def main():
             llm.register_function(name, function)
+        
+        # Check if selective MCP functions are configured
+        if config.SELECTIVE_MCP_FUNCTIONS:
```
Expected behaviour

BoraYaswanthReddy commented on app/agents/voice/automatic/services/mcp/automatic_client.py:
```diff
@@ -103,23 +108,38 @@ async def register_tools(self, llm) -> ToolsSchema:
-                logger.debug(f"Registering remote tool: {tool_name}")
+                
+                # Skip tools not in selective_functions if filtering is enabled
+                if is_selective and tool_name not in selective_functions_set:
```
That is expected behaviour

BoraYaswanthReddy commented on app/agents/voice/automatic/services/mcp/automatic_client.py:
```diff
@@ -103,23 +108,38 @@ async def register_tools(self, llm) -> ToolsSchema:
                 return ToolsSchema(standard_tools=[])
 
             raw_tools = response_dict["result"]["tools"]
+            selective_functions_set = set(selective_functions) if is_selective else None
```
ewguster_tools gets called during the initialization phase only, so this is not an issue

BoraYaswanthReddy commented on app/agents/voice/automatic/services/mcp/automatic_client.py:
```diff
@@ -103,23 +108,38 @@ async def register_tools(self, llm) -> ToolsSchema:
                 return ToolsSchema(standard_tools=[])
 
             raw_tools = response_dict["result"]["tools"]
+            selective_functions_set = set(selective_functions) if is_selective else None
```
This function gets called only once per session, so not an issue

BoraYaswanthReddy commented on app/agents/voice/automatic/__init__.py:
```diff
@@ -138,20 +153,29 @@ async def main():
             llm.register_function(name, function)
+        
+        # Check if selective MCP functions are configured
+        if config.SELECTIVE_MCP_FUNCTIONS:
```
redundant check

BoraYaswanthReddy commented on app/agents/voice/automatic/__init__.py:
```diff
@@ -138,20 +153,29 @@ async def main():
             llm.register_function(name, function)
+        
+        # Check if selective MCP functions are configured
+        if config.SELECTIVE_MCP_FUNCTIONS:
```
redundant check

BoraYaswanthReddy commented on app/agents/voice/automatic/__init__.py:
```diff
@@ -42,6 +42,21 @@
 from app.agents.voice.automatic.analytics.tracing_setup import setup_tracing
 
+def create_mcp_context(args, mode):
+    """Create a context dictionary for MCP client initialization."""
```
Not necessary as the name themselves are very descriptive

swaroopvarma1 commented on app/agents/voice/automatic/__init__.py:
```diff
@@ -42,6 +42,21 @@
 # import setup_tracing from tracing_setup.py file
 from app.agents.voice.automatic.analytics.tracing_setup import setup_tracing
 
+def create_mcp_context(args, mode):
```
move this to func to mcp/automatic/utils file

BoraYaswanthReddy replied:
That path doesnt exists
Ill move that function to the place where we currently write util functions

BoraYaswanthReddy replied:
That directory doesn't exist, so i used existing util file.
Let me know if you want me to move the function to a different place

BoraYaswanthReddy commented on app/agents/voice/automatic/__init__.py:
```diff
@@ -138,20 +139,34 @@ async def main():
             llm.register_function(name, function)
+        
+        # Check if selective MCP functions are configured
+        if config.SELECTIVE_MCP_FUNCTIONS:
```
done

BoraYaswanthReddy committed:
BZ-43226: feat: Add selective MCP server tool registration feature
- Added register_selective_tools method to MCPClient for registering specific MCP tools
- Updated __init__.py to conditionally use selective tool registration

swaroopvarma1 commented on app/agents/voice/automatic/services/mcp/automatic_client.py:
```diff
@@ -85,11 +85,11 @@ def __init__(self, server_url: str, auth_token: str, context: Dict[str, Any]):
         self._llm = llm
-        logger.info("Registering tools from custom MCP client...")
-        
+        is_selective = selective_functions is not None
```
why logger is removed? add logs for passed selective_functions
